lib/src/git.rs

```diff
@@ -488,10 +541,18 @@ pub fn fetch(
             mut_repo,
             git_repo,
             git_settings,
-            move |git_ref_name: &str| -> bool { regex.is_match(git_ref_name) },
+            move |git_ref_name: &str| -> ImportMethod {
+                if regex.is_match(git_ref_name) {
+                    ImportMethod::MergeOrResurrect
+                } else {
+                    ImportMethod::Skip
+                }
+            },
         )
     } else {
-        import_refs(mut_repo, git_repo, git_settings)
+        import_some_refs(mut_repo, git_repo, git_settings, |_| {
+            ImportMethod::MergeOrResurrect
+        })
```

Comment:
> I think it's equally annoying if forgotten branches are resurrected by jj git fetch.

I'm going to rework this a bit so that only remote-tracking branches are imported at this point.

I'm not sure whether that addresses your question at all; the question could be in regard to several things, but for most of the others I wouldn't expect it to show up at this line of code.

> It might be nice if we had a set of "tracked" remote branches. No local branch will be re-created for forgotten (or "untracked") remote branch. And to re-track such branch, user will run jj branch track.

This is a very interesting idea. One possible conclusion from this PR and all the `jj branch forget` tests I'm adding is that `jj branch forget` is a bad command to have, and some other abstraction could be better. I haven't thought your specific idea through entirely, but it seems to match what I usually want to do better.

Reply:
> I'm going to rework this a bit so that only remote-tracking branches are imported at this point.

This is now more or less done. I hoped to make `fetch` actually not touch other branches, but I had a test failure with https://github.com/ilyagr/jj/commit/tmp_onlyremotes. I couldn't immediately decide whether it's safe to change the corresponding test. Here is the portion until the failing assertion:
 https://github.com/martinvonz/jj/blob/8a8b96a448a949c50e40966ead8eac0c730ce158/lib/tests/test_git.rs#L1561-L1579

Reply:
I mean I want `jj branch forget` persist as much as possible. The current behavior isn't great, but it doesn't resurrect forgotten (i.e. uninteresting) branch so long as the remote is unchanged.

Perhaps, I would probably want `jj git fetch` that doesn't import most of the remote branches. I'd like to get remote changes, but doesn't want to maintain branches they have.

> `jj branch forget` is a bad command to have, and some other abstraction could be better.

I agree.

Reply:
> It might be nice if we had a set of "tracked" remote branches. No local branch will be re-created for forgotten (or "untracked") remote branch. And to re-track such branch, user will run jj branch track.
>
> [...]
>
> Perhaps, I would probably want jj git fetch that doesn't import most of the remote branches. I'd like to get remote changes, but doesn't want to maintain branches they have.

I've also felt this way: I don't want to logically import a given remote branch without explicitly asking for it.

Reply:
> I mean I want jj branch forget persist as much as possible. The current behavior isn't great, but it doesn't resurrect forgotten (i.e. uninteresting) branch so long as the remote is unchanged.

Yes, but that means there's no way to change your mind after you forgot the branch. That puts you into the confusing situation that @arxanas originally described. I supposed one possible answer would be "find the correct `jj branch forget` command in `jj op log` and `jj undo` it", but that doesn't sound great.

Reply:
> there's no way to change your mind after you forgot the branch

Yeah, that's the problem of `jj branch forget`, I think. If `BranchTarget::local_target` were tristate `Deleted|Forgot|Some(RefTarget)` for example, remote branch wouldn't have to be deleted by `jj branch forget`, and we wouldn't need special case in `git::import_refs()`. I think it will also make `git.auto-local-branch = false` less scary.

https://github.com/martinvonz/jj/blob/dee5dff20dbe834fad799d42b2354f0068b03928/lib/src/op_store.rs#L175-L183

FWIW, I'm not sure if it's good idea to import remote branches into single local namespace. Naming is arbitrary, so my "hack" branch might be unrelated to their "hack". And I wouldn't wanna be bothered to resolve branch conflicts on every fetch.

Reply:
> Yeah, that's the problem of jj branch forget, I think. If BranchTarget::local_target were tristate Deleted|Forgot|Some(RefTarget) for example, remote branch wouldn't have to be deleted by jj branch forget, and we wouldn't need special case in git::import_refs(). 

I agree, this is very similar to your original `track` suggestion and what I wrote below in https://github.com/martinvonz/jj/pull/1714#issuecomment-1599789564. It will take a bit of thought about the UI and fetch behavior to implement that, as I explained there.

> FWIW, I'm not sure if it's good idea to import remote branches into single local namespace. Naming is arbitrary, so my "hack" branch might be unrelated to their "hack". And I wouldn't wanna be bothered to resolve branch conflicts on every fetch.

I, personally, never had a problem with that, but @martinvonz regularly suggests that he's thinking about changing the branch model to something git-like. I think this is pretty much what he means.

Reply:
(I updated my previous comment; I think I didn't entirely follow your idea initially).

Reply:
> I, personally, never had a problem with that, but @martinvonz regularly suggests that he's thinking about changing the branch model to something git-like. I think this is pretty much what he means.

I don't run into any problems with that kind of conflict either. I just mean that I'm open to changing the model if we think the current model is causing too many problems (I'm thinking mostly of issues like the one fixed in this PR).

But as I said on Discord, before we change to a more git-like model, I think we should make sure that we're solving the right problem. For example, if people often find a need to forget local branches because they find uninteresting branches distracting, maybe we can solve that by instead by making uninteresting branches less visible in different places. Maybe we define an "uninteresting branch" as one where the local target matches some remote's target.

Reply:
> Maybe we define an "uninteresting branch" as one where the local target matches some remote's target.

This wouldn't work great for me with the Github fork workflow: my branches regularly match branches on my fork, and are still interesting to me. So, I think that if we want to support Github fork workflow and use this kind of criterion, we'd want to define one of the forks as the "upstream" fork.

Throughout this PR, we have a solidifying proposal to have people list which forks they'd like to forget about. I'm not sure if we also (or instead?) want some in-between state ("I'm not interested in this branch, but I don't want to forget it entirely"). @yuja's proposal above about "... if `BranchTarget::local_target` were tristate `Deleted|Forgot|Some(RefTarget)` for example ..." could be adapted into something like that.

Reply:
Actually I don't use (named) branches a lot. I run `jj branch list` only when `jj status` complained about something conflicted blah blah... (maybe because I did `jj abandon` without deleting local branches?) So my `jj branch forget` always means "don't bother me again."

Oh, my problem can be solved if `jj abandon` doesn't move a branch backwards? But we'll need to show some kind of "orphaned" state if a branch isn't reachable from the current heads.

> On second thought, this is a bit different since you are also suggesting not deleting remote-tracking branches when forgetting

Yes. My point is that the current behavior is peculiar because we have invisible state about remote refs, so I want to make them visible. If `jj branch forget` really "forget"s, it should remove remote-tracking refs as well (on `jj git export` I suppose.) If it's "untrack" command, we'll need to remember untracked branches forever so user can re-track them when needed.

> I, personally, never had a problem with that

I don't have either. I just guessed this could happen from the list of branches in my jj repo.

Reply:
> Oh, my problem can be solved if jj abandon doesn't move a branch backwards? But we'll need to show some kind of "orphaned" state if a branch isn't reachable from the current heads.

This is an interesting idea to think about. At the moment, my only observation is that remote-tracking branches are already often at orphaned commits. 

> , it should remove remote-tracking refs as well (on jj git export I suppose.)

I think that would be reasonable in normal jj repo. In colocated repos, the git_ref tacking the local git repo's branch must **not** be removed by `jj branch forget`, but it **will** be removed by the subsequent `jj git export` as it deletes the local git repo's branch (which I'm pretty sure is the desired behavior).

> If it's "untrack" command, we'll need to remember untracked branches forever so user can re-track them when needed.

We're both OK with this in principle, correct?

Reply:
> The git_ref tacking the local git repo's branch must **not** be removed by `jj branch forget`, but it **will** be removed by the subsequent `jj git export` as it deletes the local git repo's branch

I mean `.git/refs/remotes/<remote>/<branch>` should also be removed (on git export) because that's what forgetting remote branches supposed to do.
 
> > If it's "untrack" command, we'll need to remember untracked branches forever so user can re-track them when needed.
> 
> We're both OK with this in principle, correct?

I think yes, in the current jj's branch model. If we decide to switched to git-like branch model, things will be different.

Reply:
I think this is resolved as far as this PR goes, but feel free to un-resolve it if I rushed.

> I mean .git/refs/remotes/<remote>/<branch> should also be removed (on git export) because that's what forgetting remote branches supposed to do.

I'd rather create a list of forgotten branches and have them be ignored. Currently, only `jj git push` is allowed to modify these refs, and I think that's a nice property to preserve unless we have to break it.

> If we decide to switched to git-like branch model, things will be different.

I wouldn't be overly surprised if there's a need for switching to the git model eventually, but maybe not yet or maybe a very compelling reason for it won't come up? I'm a bit scared of thinking about making `git import/export/fetch/push` work with a change that big.

Reply:
> > I mean .git/refs/remotes// should also be removed (on git export) because that's what forgetting remote branches supposed to do.
> 
> I'd rather create a list of forgotten branches and have them be ignored.

Well, I would do that if "forget" is persistent (which I said I would prefer.) For the current "documented" behavior, I think forgetting `.git/refs/remotes` (as well as our `git_refs`) is correct.

Anyway, I don't know for what purpose people use `jj branch forget`. Mine appears to be just a workaround for accidental branch move on `jj abandon`, which seems not a valid usecase. And I lean towards not applying ad-hoc fix until we determine the right model.

Reply:
OK, I assume you don't want this resolved, so I un-resolved it.

> Anyway, I don't know for what purpose people use jj branch forget. 

I mainly used it when I accidentally fetched too many branches (e.g. accidentally did `jj git fetch --remote upstream` instead of `jj git fetch --remote upstream --branch main`). For that, the current documented behavior is what I'd want.

>  I lean towards not applying ad-hoc fix until we determine the right model.

I was just trying to write up an outline of what I thought a decent model would be. I'll post it as a feature request and give a link.

As for what gets fixed first, I'm unsure what the best option is. However, since you feel it's early to merge this PR, I think I'll put its first 3 commits in a separate PR. I also want to fix the bug where `jj branch forget` refuses to work on deleted branches, I don't think it's blocked by this discussion.

Reply:
Here's my write-up: https://github.com/martinvonz/jj/issues/1734

Reply:
> I mainly used it when I accidentally fetched too many branches

Isn't `undo` a better option in that case? I guess `undo` wouldn't work at least for colocated repo, but I think it's the same kind of problem we've discussed in the other `jj undo` thread.

> As for what gets fixed/merged first, I'm unsure what the best option is.

Does this `forget` + `fetch` issue need an immediate fix? I feel it's not serious bug, but I might be wrong.

> Here's my write-up: https://github.com/martinvonz/jj/issues/1734

Thanks!

Reply:
> Isn't undo a better option in that case? I guess undo wouldn't work at least for colocated repo, but I think it's the same kind of problem we've discussed in the other jj undo thread.

Sometimes a bit of time has passed and it's hard to find in the op-log.

Sometimes it might be easier to fetch all branches and then forget some globs than to fetch some globs.

More generally, I think that, as far as possible, anything `undo` can do should be doable by regular commands for similar reasons (at least, that I can think of now): it may be hard to search the oplog or you might want to partially undo something. I'm not sure how important this principle is to me, but this situation feels like it should apply.

> I guess undo wouldn't work at least for colocated repo

Aargh. At the moment, I have more emotions than words for this.

I actually think it might work and we might even have a test for it, but I need to double-check. If not, I should add a test to #1700 to see what happens, at least.

> Does this forget + fetch issue need an immediate fix? I feel it's not serious bug, but I might be wrong.

I felt like the situation from Discord that I put in the summary is quite a confusing situation for a user to end up in. I find it hard to weigh that against you (and potentially other people) who may be relying on the current behavior.

Reply:
Another use-case for `branch forget` I just came upon: you did a `jj branch delete` and realized that you don't want to destroy the branch on push. I just remembered it after looking at #1537.

Reply:
> I think that, as far as possible, anything undo can do should be doable by regular commands

Yes, I agree about that. It might be a maintenance command like `jj git forget-refs` or option to ignore known git_refs state, but I don't know.